File: step_definitions/Verify User Can View Past Orders and Order Details.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.OrderHistoryPage;

public class OrderHistorySteps extends OrderHistoryPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user has valid credentials")
    public void theUserHasValidCredentials() {
        // Implement login logic here
    }

    @When("the user logs in with User ID {string}")
    public void theUserLogsInWithUserID(String userID) {
        // Implement login logic here
    }

    @Then("the user is redirected to the dashboard")
    public void theUserIsRedirectedToTheDashboard() {
        // Implement redirection verification logic here
    }

    @Given("the user is on the dashboard")
    public void theUserIsOnTheDashboard() {
        // Implement dashboard navigation logic here
    }

    @When("the user navigates to the {string} section")
    public void theUserNavigatesToTheSection(String section) {
        navigateToOrderHistory();
    }

    @Then("the {string} section is displayed with a list of past orders")
    public void theSectionIsDisplayedWithAListOfPastOrders(String section) {
        verifyOrderHistoryDisplayed();
    }

    @Given("the user is on the {string} section")
    public void theUserIsOnTheSection(String section) {
        // Implement logic to ensure user is on the specified section
    }

    @When("the user selects an order with Order ID {string}")
    public void theUserSelectsAnOrderWithOrderID(String orderID) {
        selectOrder(orderID);
    }

    @Then("the order details are displayed including product names, quantities, prices, and total cost")
    public void theOrderDetailsAreDisplayedIncludingProductNamesQuantitiesPricesAndTotalCost() {
        verifyOrderDetailsDisplayed();
    }

    @Then("the order date {string} and status match the expected values")
    public void theOrderDateAndStatusMatchTheExpectedValues(String expectedDate) {
        // Implement logic to verify order date and status
    }

    @Then("the product details such as name {string}, quantity {string}, and price are accurate")
    public void theProductDetailsSuchAsNameQuantityAndPriceAreAccurate(String productName, String quantity) {
        // Implement logic to verify product details
    }

    @Then("the total amount {string} matches the sum of product prices and quantities")
    public void theTotalAmountMatchesTheSumOfProductPricesAndQuantities(String expectedTotal) {
        // Implement logic to verify total amount
    }

    @Then("any applicable discounts or promotions are accurately applied to the total amount")
    public void anyApplicableDiscountsOrPromotionsAreAccuratelyAppliedToTheTotalAmount() {
        // Implement logic to verify discounts and promotions
    }

    @Then("the shipping address and billing address are correct and match the user's information")
    public void theShippingAddressAndBillingAddressAreCorrectAndMatchTheUsersInformation() {
        // Implement logic to verify addresses
    }

    @Then("the payment method matches the method used during checkout")
    public void thePaymentMethodMatchesTheMethodUsedDuringCheckout() {
        // Implement logic to verify payment method
    }

    @When("the user attempts to download the order invoice or receipt")
    public void theUserAttemptsToDownloadTheOrderInvoiceOrReceipt() {
        // Implement logic to download invoice
    }

    @Then("the invoice or receipt is successfully downloaded and contains accurate order information")
    public void theInvoiceOrReceiptIsSuccessfullyDownloadedAndContainsAccurateOrderInformation() {
        // Implement logic to verify downloaded invoice
    }

    @When("the user filters orders by a specified date range")
    public void theUserFiltersOrdersByASpecifiedDateRange() {
        // Implement logic to filter orders by date range
    }

    @Then("orders are filtered and displayed according to the specified date range")
    public void ordersAreFilteredAndDisplayedAccordingToTheSpecifiedDateRange() {
        // Implement logic to verify filtered orders
    }

    @When("the user sorts orders by status")
    public void theUserSortsOrdersByStatus() {
        // Implement logic to sort orders by status
    }

    @Then("orders are sorted correctly based on status")
    public void ordersAreSortedCorrectlyBasedOnStatus() {
        // Implement logic to verify sorted orders
    }

    @When("there are large numbers of orders")
    public void thereAreLargeNumbersOfOrders() {
        // Implement logic to handle large number of orders
    }

    @Then("pagination is implemented and works correctly for navigating through orders")
    public void paginationIsImplementedAndWorksCorrectlyForNavigatingThroughOrders() {
        // Implement logic to verify pagination
    }

    @Given("the user is on the order details page")
    public void theUserIsOnTheOrderDetailsPage() {
        // Implement logic to ensure user is on order details page
    }

    @When("the user navigates back to the dashboard")
    public void theUserNavigatesBackToTheDashboard() {
        // Implement logic to navigate back to dashboard
    }

    @Then("the user is redirected to the dashboard without any issues")
    public void theUserIsRedirectedToTheDashboardWithoutAnyIssues() {
        // Implement logic to verify redirection
    }

    @When("the user logs out from the account")
    public void theUserLogsOutFromTheAccount() {
        // Implement logout logic
    }

    @Then("the user is successfully logged out and redirected to the login page")
    public void theUserIsSuccessfullyLoggedOutAndRedirectedToTheLoginPage() {
        // Implement logic to verify logout
    }
}

File: step_definitions/Verify User Can View Past Orders and Order Details.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import DashboardPage from "../pages/dashboardPage";
import OrderHistoryPage from "../pages/orderHistoryPage";
import OrderDetailsPage from "../pages/orderDetailsPage";

Given("the user has valid credentials", () => {
 // Assume credentials are stored in environment variables
});

When("the user logs in with User ID {string}", (userId) => {
 cy.login(userId, Cypress.env('password'));
});

Then("the user is redirected to the dashboard", () => {
 cy.url().should("include", "/dashboard");
});

Given("the user is on the dashboard", () => {
 DashboardPage.visit();
});

When("the user navigates to the {string} section", (section) => {
 DashboardPage.goToSection(section);
});

Then("the {string} section is displayed with a list of past orders", (section) => {
 OrderHistoryPage.verifySectionDisplayed(section);
});

Given("the user is on the {string} section", (section) => {
 OrderHistoryPage.visit();
});

When("the user selects an order with Order ID {string}", (orderId) => {
 OrderHistoryPage.selectOrder(orderId);
});

Then("the order details are displayed including product names, quantities, prices, and total cost", () => {
 OrderDetailsPage.verifyOrderDetails();
});

Given("the order details are displayed", () => {
 OrderDetailsPage.verifyOrderDetailsDisplayed();
});

Then("the order date {string} and status match the expected values", (date) => {
 OrderDetailsPage.verifyOrderDateAndStatus(date);
});

Then("the product details such as name {string}, quantity {string}, and price are accurate", (productName, quantity) => {
 OrderDetailsPage.verifyProductDetails(productName, quantity);
});

Then("the total amount {string} matches the sum of product prices and quantities", (totalAmount) => {
 OrderDetailsPage.verifyTotalAmount(totalAmount);
});

Then("any applicable discounts or promotions are accurately applied to the total amount", () => {
 OrderDetailsPage.verifyDiscountsAndPromotions();
});

Then("the shipping address and billing address are correct and match the user's information", () => {
 OrderDetailsPage.verifyAddresses();
});

Then("the payment method matches the method used during checkout", () => {
 OrderDetailsPage.verifyPaymentMethod();
});

When("the user attempts to download the order invoice or receipt", () => {
 OrderDetailsPage.downloadInvoice();
});

Then("the invoice or receipt is successfully downloaded and contains accurate order information", () => {
 OrderDetailsPage.verifyInvoiceDownloaded();
});

When("the user filters orders by a specified date range", () => {
 OrderHistoryPage.filterOrdersByDateRange();
});

Then("orders are filtered and displayed according to the specified date range", () => {
 OrderHistoryPage.verifyFilteredOrders();
});

When("the user sorts orders by status", () => {
 OrderHistoryPage.sortOrdersByStatus();
});

Then("orders are sorted correctly based on status", () => {
 OrderHistoryPage.verifySortedOrders();
});

When("there are large numbers of orders", () => {
 OrderHistoryPage.checkLargeNumberOfOrders();
});

Then("pagination is implemented and works correctly for navigating through orders", () => {
 OrderHistoryPage.verifyPagination();
});

Given("the user is on the order details page", () => {
 OrderDetailsPage.visit();
});

When("the user navigates back to the dashboard", () => {
 OrderDetailsPage.navigateBackToDashboard();
});

Then("the user is redirected to the dashboard without any issues", () => {
 cy.url().should("include", "/dashboard");
});

When("the user logs out from the account", () => {
 DashboardPage.logout();
});

Then("the user is successfully logged out and redirected to the login page", () => {
 cy.url().should("include", "/login");
});